### ISSUE TITLE
Request to divide into two separate: IntelliSense and Debugging

### ISSUE THREAD

**Issue by dmyger** [microsoft/vscode-cpptools#2656]
Is it possible split current extension in two separate: one for language service another for debugging?

As for me, there is a small issue: no matter how hard I tried, I couldn't make C++ IntelliSense works as I want for my projects, so I just don't use all features of the IntelliSense. I have to set 'disable' settings wherever it's possible, to avoid interfering with cquery, which satisfies me completely. But actually it's possible only if disable extension at all.

On other hand, your extension has very good features for debugging and I prefer use it. So I need to enable back the extension before debugging. This case is not very convenient to use.

As for me, it seems that would be more useful to have two extensions with clearly separated functionality. What do you think about it?

**Comment by pieandcakes** [comment]
@bobbrow Do we have a way to disable Intellisense for @dmyger  ?

**Comment by pieandcakes** [comment]
@dmyger We are trying to provide a complete experience, especially for new users. If you are having issues with Intellisense and would like to report issues, they can look into your scenario and see if they can resolve the problems you are running into.

**Comment by sean-mcmanus** [comment]
We already have a "Disabled" setting for "C_Cpp.intelliSenseEngine". Does that work for you?

**Comment by dmyger** [comment]
I have following disables:
```json
"C_Cpp.intelliSenseEngine": "Disabled",
"C_Cpp.autocomplete": "Disabled",
"C_Cpp.formatting": "Disabled",
"C_Cpp.errorSquiggles": "Disabled",
"C_Cpp.intelliSenseEngineFallback": "Disabled",
```
And when I click mouse on function with Ctrl pressed (`Go to Definition`) I get two offers one from cquery another one from "disabled" IntelliSense, both are in the same file and the same line. But, more worst situation when I try do it with any variable I get a lot of definitions wherever it find variables with same name, instead of completely accurate navigation to the desired definition of a single variable as it works when the IntelliSense extension disabled at all.

**Comment by bobbrow** [comment]
Changing the `intelliSenseEngine` setting requires you to reload the VS Code window to take effect.  Did you reload the window and the language service is still running?

**Comment by dmyger** [comment]
Of course, I'm reloading a window and doing full restart of application, too.
But I don't said that _"language service is still running"_, I said that some of its features are keep working.

**Comment by sean-mcmanus** [comment]
What features are still working? The only thing that should linger around are the menu commands that are "statically defined" in the package.json. I'm not able to repro Go to Definition giving any results when the intelliSenseEngine is set to Disabled.

**Comment by dmyger** [comment]
> What features are still working?
>> And when I click mouse on function with Ctrl pressed (Go to Definition) I get two offers one from cquery another one from "disabled" IntelliSense, both are in the same file and the same line. But, more worst situation when I try do it with any variable I get a lot of definitions wherever it find variables with same name,

I'm not sure that it will be easy to reproduce, cause of non trivial configuration of extensions and the project.
To reproduce you have to install and configure `cquery` to use _compile_commands.json_ from the `CMake Tools` extension. Then you need open, configure and build a project with the CMake. And only after it you can try navigating inside the code, to see the issue that VS Code offers doubled functions: one from cquery and another one from IntelliSense, despite last was disabled.

That's why I had asked to split the extension. In real life more suitable have plenty of small solution than a few but large.

If you want, I can reproduce it and send you any logs or what ever else what I can collect from my instance of your application. Only say me what to do.

**Comment by sean-mcmanus** [comment]
How do you know the double results are from our extension? Do you see a Microsoft.VSCode.CPP.Extension process running? The code that does the disabling is at https://github.com/Microsoft/vscode-cpptools/blob/release/Extension/src/main.ts line 279. It checks for the disabled setting and then returns before it activates our language server -- I don't see any way this could be buggy.

**Comment by dmyger** [comment]
Ok, lets try again. As I said about it early, when I'm disabling an IntelliSense as the entire extension, and trying navigate to function or variable definition I get completely accurate navigation to the desired definition of a single variable **without any extra screens**, due to work of `cquery`. But if I turn on the `Microsoft.VSCode.CPP.Extension` with disabling only IntelliSense part of it (as it was mentioned above) and trying navigate to definition I'll get a lot of wrong definitions wherever it was find the same names, see attached screen shot:
![screenshot](https://user-images.githubusercontent.com/7373914/47165919-b895fc80-d303-11e8-9569-fd3a1690f729.png)
If check the same with unique name of function/variable I'll get two records in this popup.

Of course, I can't see nor what lead to such issue, nor  a _"Microsoft.VSCode.CPP.Extension process running"_, but I made a conclusion if as my settings changed only by turning on one extensions, then such extension is the reason of such behavior.

**Comment by bobbrow** [comment]
Can you [Enable Logging](https://github.com/Microsoft/vscode-cpptools/blob/master/Documentation/LanguageServer/Enabling%20logging.md) and see what our extension is printing to the Output window?

**Comment by dmyger** [comment]
Log after enable extension till "go to definition" [cpp_ext.log](https://github.com/Microsoft/vscode-cpptools/files/2492728/cpp_ext.log)
With workspace [settings](https://github.com/Microsoft/vscode-cpptools/issues/2656#issuecomment-429810020)

**Comment by bobbrow** [comment]
Fascinating.  It looks like the registration of a custom configuration provider is overriding our activation policy.  I'll investigate.  Thanks for the log!

**Comment by bobbrow** [comment]
The fix will be available in 0.19.1.  Thanks again for the logs.

**Comment by sean-mcmanus** [comment]
The Disable intelliSenseEngine setting should work with CMake Tools now with https://github.com/Microsoft/vscode-cpptools/releases/tag/0.19.1-insiders3 , so splitting the language service and debugger into separate extensions shouldn't be needed.
